### PR TITLE
Feature - grouping delegate

### PR DESCRIPTION
**Pull request by joefitter**
This PR adds an optional delegate field to a grouping extension. The delegate is authorised to mint new assets into the group.

The main use case for this change is a program controlling either a mint (candy machine) or a conversion from another asset type.

* Added delegate field to grouping extension
* Updated group processor to check grouping delegate if present
* Updated create instruction to accept an optional grouping_delegate signer

Todo: Build changes into interface program - this isn't rebuilt when running programs:build, so the generation never runs and tests fail

Also perhaps worth considering not reinitialising the data when the max_size is written as there are likely cases where this may need to be updated post creation.

### CODE REVIEW

File: programs/asset/interface/src/interface.rs [hunk 1]
```diff
@@ -29,7 +29,8 @@ pub enum Interface {
     #[account(2, name="owner", desc = "The owner of the asset")]
     #[account(3, optional, writable, name="group", desc = "Asset account of the group")]
     #[account(4, optional, signer, writable, name="payer", desc = "The account paying for the storage fees")]
-    #[account(5, optional, name="system_program", desc = "The system program")]
+    #[account(5, optional, signer, name="grouping_delegate", desc = "The delegate authority for minting assets into a group")]
```
Comment (febo) [inline programs/asset/interface/src/interface.rs:32]: I think we can name this as `group_authority` since it technically can either be the actual authority of the group or delegate.

Nitpick: can we also move it to be after the group account>

File: programs/asset/program/src/instruction.rs [hunk 2]
```diff
@@ -28,7 +28,8 @@ pub enum Instruction {
     #[account(2, name="owner", desc = "The owner of the asset")]
     #[account(3, optional, writable, name="group", desc = "Asset account of the group")]
     #[account(4, optional, signer, writable, name="payer", desc = "The account paying for the storage fees")]
-    #[account(5, optional, name="system_program", desc = "The system program")]
+    #[account(5, optional, signer, name="grouping_delegate", desc = "The delegate authority for minting assets into a group")]
```
Comment (febo) [inline programs/asset/program/src/instruction.rs:31]: Same as the comment on the `instruction.rs`.

File: programs/asset/program/src/processor/create.rs [hunk 3]
```diff
@@ -26,7 +26,8 @@ use crate::{
 ///   2. `[]` owner
 ///   3. `[writable, optional]` group
 ///   4. `[writable, signer, optional]` payer
-///   5. `[optional]` system_program
+///   5. `[signer, optional]` grouping_delegate
```
Comment (febo) [inline programs/asset/program/src/processor/create.rs:29]: Update this comment after the change on the account name.

File: programs/asset/program/src/processor/create.rs [hunk 4]
```diff
@@ -222,14 +223,17 @@ pub fn process_create(
...
+
         super::group::process_group(
             program_id,
             Context {
                 accounts: GroupAccounts {
-                    authority: ctx.accounts.authority,
+                    authority: ctx
+                        .accounts
+                        .grouping_delegate
+                        .unwrap_or(ctx.accounts.authority),
```
Comment (febo) [inline programs/asset/program/src/processor/create.rs:236]: Nice.

File: programs/asset/program/src/processor/group.rs [hunk 5]
```diff
@@ -91,6 +77,26 @@ pub fn process_group(program_id: &Pubkey, ctx: Context<GroupAccounts>) -> Progra
         );
     };
 
+    if let Some(delegate) = grouping.delegate.value() {
+        require!(
+            delegate == &NullablePubkey::new(*ctx.accounts.authority.key),
```
Comment (febo) [inline programs/asset/program/src/processor/group.rs:82]: I think you can write this test as:
```rust
*delegate == ctx.accounts.authority.key
```

File: programs/asset/program/src/processor/group.rs [hunk 6]
```diff
@@ -60,29 +60,15 @@ pub fn process_group(program_id: &Pubkey, ctx: Context<GroupAccounts>) -> Progra
...
-
     require!(
         asset.group.value().is_none(),
         AssetError::AlreadyInGroup,
         "asset"
     );
 
-    let group = Asset::load_mut(&mut group_data);
-
-    require!(
```
Comment (febo) [inline programs/asset/program/src/processor/group.rs:77]: I think we still need to assert that the `asset.authority` matches the `group.authority` (move the comments from line 59 to here as well).

File: programs/asset/program/src/processor/group.rs [hunk 7]
```diff
@@ -91,6 +77,26 @@ pub fn process_group(program_id: &Pubkey, ctx: Context<GroupAccounts>) -> Progra
         );
     };
 
+    if let Some(delegate) = grouping.delegate.value() {
```
Comment (febo) [inline programs/asset/program/src/processor/group.rs:80]: I think we need to invert this condition. This will prevent the actual authority to add an asset to a group when there is a delegate since it will try to match against the delegate.

I think we need to first check if `*ctx.accounts.authority.key == group.authority`; if this fails, then we look for a delegate.

If both conditions fail, then the authority is invalid.

File: programs/asset/program/src/processor/group.rs [hunk 8]
```diff
@@ -60,29 +60,15 @@ pub fn process_group(program_id: &Pubkey, ctx: Context<GroupAccounts>) -> Progra
...
-    require!(
-        group.authority == *ctx.accounts.authority.key,
-        AssetError::InvalidAuthority,
-        "group authority mismatch"
-    );
-
     // group size validation
-
-    let grouping = if let Some(grouping) = Asset::get_mut::<GroupingMut>(&mut group_data) {
+    let group = Asset::load(&group_data);
```
Comment (febo) [inline programs/asset/program/src/processor/group.rs:70]: I think you can revert this change (and the one on the following line).

File: programs/asset/program/src/processor/group.rs [hunk 9]
```diff
@@ -100,6 +106,8 @@ pub fn process_group(program_id: &Pubkey, ctx: Context<GroupAccounts>) -> Progra
     }
 
     // assign the group to asset and increment the group size
+    let grouping = Asset::get_mut::<GroupingMut>(&mut group_data).unwrap();
```
Comment (febo) [inline programs/asset/program/src/processor/group.rs:109]: I think you can revert this change and the one on the following line.

**Review by febo — CHANGES_REQUESTED** [review]
Good stuff! There are small little things on the logic for using the delegate. We should also do the same logic on the `ungroup` instruction, i.e., allow the grouping delegate to ungroup an asset from a group.

**Comment by febo** [comment]
You can also run `pnpm format:fix` on `./clients/js/asset` folder to fix the lint error.

File: programs/asset/program/src/processor/group.rs (hunk 5 above)
Comment (joefitter) [reply programs/asset/program/src/processor/group.rs:82]: It complains about a mismatch between`NullablePubkey` and `&Pubkey`, however this works:
```rust
*delegate == ctx.accounts.authority.key.into()
```

File: programs/asset/program/src/processor/group.rs (hunk 6 above)
Comment (joefitter) [reply programs/asset/program/src/processor/group.rs:77]: yes, good shout

File: programs/asset/program/src/processor/group.rs (hunk 9 above)
Comment (joefitter) [reply programs/asset/program/src/processor/group.rs:109]: I was wrestling with the borrow checker, I'll wrestle harder :)

**Comment by joefitter** [comment]
Thanks for your comments! All addressed pending rebuilding the interface program

**Comment by febo** [comment]
> Thanks for your comments! All addressed pending rebuilding the interface program

Amazing! I will rebuild the interface on a separate PR after we merge this one.

Can you please run the lint on the JS client as well?
```
pnpm lint:fix
```

**Review by febo — APPROVED** [review]
Looks great, thank you!

CI is failing because the interface is not updated – I will do it on a separate PR.
